### ISSUE TITLE
Freeplane versions after v1.10.5 pre 3 are much slower when displaying a map with conditional styles

### ISSUE THREAD

**Issue by EdoFro** [freeplane/freeplane#831]
**Describe the bug**
Freeplane versions after v1.10.5 pre 3 are much slower when displaying a map with conditional styles

**To Reproduce**
1. Open map TestMap.mm in Freeplane
2. Select root child node "Compartida Infraestructura"
3. Execute menu command "Unfold all"
4. Wait until the UI refreshes
5. Install another version from Freeplane and repeat the test

**Expected behavior**
new versions should be as fast as v1.10.5 pre3

**Mind map files**
I will private mail you the map. I tried before to anonimize the map to attach it here but the bug was not as big.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: windows
 - Version: 10
 - Java Version: 15.0.2

**Additional context**
I used the Portable Apps version of Freeplane to easily instal the different versions of Freeplane.
It uses portable java: AdoptOpenJDK JRE "15.0.2+7"
[JAVArelease.zip](https://github.com/freeplane/freeplane/files/10036321/JAVArelease.zip)
each time I installed a different version of Freeplane I changed the **freeplane.l4j.ini files** to `-Xmx2G`

my properties file: 
[auto.properties.zip](https://github.com/freeplane/freeplane/files/10036339/auto.properties.zip)

**Comment by EdoFro** [comment]
Some of the conditional styles rules need an addon to be instaled.
[groovyNode-v0.0.2.addon.zip](https://github.com/freeplane/freeplane/files/10036409/groovyNode-v0.0.2.addon.zip)

it can be easily uninstalled afterwards.

**Comment by EdoFro** [comment]
as mentioned in this [post](https://github.com/freeplane/freeplane/discussions/725#discussioncomment-4172137), I made multiple test with the TestMap.mm in the different Freeplane versions.

Each test was activating certain conditional rules. For each time, I loaded the mindmap, applied the test and closed the mindmap whitout saving the changes. This way each test was similar and comparable to the other.

These are the conditional rules:
![image](https://user-images.githubusercontent.com/35700575/202582066-0689f5b4-ccf1-4e15-81da-2a38b52ac6d9.png)

## Here its the table of results:


|Test|Conditional Styles|FP v1.10.5 pre3|FP v1.10.5 pre5|FP v1.10.5 stable|FP v1.11.1 alpha pre 2|
|--|--|--:|--:|--:|--:|
|1|All (but "For any descendant..")|1,8|60,8|52,1|16,6|
|2|None|0,7|1,2|0,8|0,6|
|3|No script filters|0,9|3,5|3,1|1,7|
|4|No `(node.findAll() - node).any{...` filters|1,2|6|5,5|3,1|
|5|No groovyNode filters|1,3|19,1|16,4|6,9|
|6|only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: Script filter (node.fidAll() ….|1,1|31,2|28,9|8,4|
|7|only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: For any descendant ….|2,3|289,5|291,3|104,8|
|8|All (but "For any descendant..") & without MapOverview|1,5|48,3|46,4|16,6|

## And these are the test that I did:

### All (but "For any descendant..")
![image](https://user-images.githubusercontent.com/35700575/202582066-0689f5b4-ccf1-4e15-81da-2a38b52ac6d9.png)

### None
![image](https://user-images.githubusercontent.com/35700575/202583825-75ceb193-0585-42ef-9334-cdf393779207.png)

### No script filters
![image](https://user-images.githubusercontent.com/35700575/202584466-63ce7a0c-688e-417b-91bc-1def50ed8e32.png)

### No `(node.findAll() - node).any{...` filters
![image](https://user-images.githubusercontent.com/35700575/202583923-09826240-f6d3-43cf-aa2d-aec6755d9fdb.png)

### No groovyNode filters
![image](https://user-images.githubusercontent.com/35700575/202584008-796a0420-cf0b-4dc6-a6c5-e358218b1beb.png)

### only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: Script filter (node.fidAll() ….
![image](https://user-images.githubusercontent.com/35700575/202584162-df689984-25e6-46c3-adda-9dabe90243f6.png)

### only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: For any descendant ….
![image](https://user-images.githubusercontent.com/35700575/202584241-35bfa979-e40a-4206-90a2-a6fd637c400d.png)

**Comment by EdoFro** [comment]
To get the time duration in seconds, I made a script that displays a dialog. 
(here is the file: [FoldUnfoldElapsedTimeTest.zip](https://github.com/freeplane/freeplane/files/10036586/FoldUnfoldElapsedTimeTest.zip))

![image](https://user-images.githubusercontent.com/35700575/202585557-c839fd80-9d07-456c-baab-7b216ebb19e2.png)

To use it:
1. you have to select the branch node you want
2. click the button 1
3. right after that click the button 2 as soon as possible (before it refreshes)
4. in the text area, it will show the duration in seconds it took. (it will also copy that value to the clipboard)


The first button executes a fold and unfold all
the second button gets the time lapse since the first button was clicked, but it can only execute after the UI has repainted itself. This way it gets the time duration

Here is the script:

```groovy
// @ExecutionModes({ON_SINGLE_NODE="/main_menu/Scripts"})

import groovy.swing.SwingBuilder
import java.awt.*

def swingBuilder = new SwingBuilder()

def myTextArea

def action1 ={
    def start = new Date().time
    println start
    node['t_ini_temp'] = start
    menuUtils.executeMenuItems(['FoldAllAction','UnfoldAllAction'])
}

def action2 ={
    def fin = new Date().time
    println fin
    def lapse = ((fin - node['t_ini_temp'].to.num0)/1000).round(1)
    println lapse
    myTextArea.text = "  $lapse seconds"
    println ''
    textUtils.copyToClipboard(lapse.toString())
    //node['t_ini_temp'] = null
}


def buttonPanel = {
    swingBuilder.panel(
        layout : new GridLayout(0,1)
    ) {
        button(text : '1. start test',             actionPerformed : action1)
        button(text : '2. get duration time', actionPerformed : action2)
   }
}



def dialogo = swingBuilder.dialog(
                title              : 'Time Duration Test',
                name               : 'durationTest',
                modal              : false,
                locationRelativeTo : ui.frame,
                layout             : new GridLayout(0,1),
                owner              : ui.frame,
                //defaultCloseOperation: JFrame.DISPOSE_ON_CLOSE,
                show               : true,
                pack               : true
            ) {
                buttonPanel()
                myTextArea = textArea(id:'texto', lineWrap: "true", wrapStyleWord: "true", columns:25, rows:4)
            }

return "ready"

```

**Comment by dpolivaev** [comment]
It is a complicated issue where multiple pieces can be improved.
Can you please test preview 1.10.6_01 ? 
It contains the first improvement, but it can influence how script conditions generally work, and therefore I need you to test it before I can make other improvements.

**Comment by EdoFro** [comment]
> Here its the table of results:

I repeated the same tests with the new versions: **1.10.6_01** and **1.11.1_03**

I included the results in the table:

|Test|Conditional Styles|1.10.5 p3|1.10.5 p5|1.10.5 stable|1.11.1 alpha pre 2|1.10.6_01|1.11.1_03|
|:--:|--|--:|--:|--:|--:|--:|--:|
|1|All (but "For any descendant..")|1,8|60,8|52,1|16,6|3,6|1,8|
|2|None|0,7|1,2|0,8|0,6|0,5|0,7|
|3|No script filters|0,9|3,5|3,1|1,7|3|0,8|
|4|No `(node.findAll() - node).any{...` filters|1,2|6|5,5|3,1|3,1|1,1|
|5|No groovyNode filters|1,3|19,1|16,4|6,9|3,2|1,4|
|6|only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: Script filter (node.fidAll() ….|1,1|31,2|28,9|8,4|0,6|0,8|
|7|only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: For any descendant ….|2,3|289,5|291,3|104,8|2,3|0,6|
|8|All (but "For any descendant..") & without MapOverview|1,5|48,3|46,4|16,6|3,9|1|

As you can see speed was greatly improved. thanks!

In the following days, I will test if the changes caused any unwanted side-effects.

I will report the details of any possible issue here

**Comment by EdoFro** [comment]
I decided to use v1.10.6_01 to test for unwanted side-effects.

I have played with conditional formating and, until now, I haven't notice any problem.

A nice side effect is that the note panel opens faster than before.

**Comment by EdoFro** [comment]
First side effect:
When creating a new node with a script, the 'conditional formating script filter' rules are not applied to the new node.

The new node will not **show** its correct styles:
- it may not show a conditional style that its assigned to it
- it may show a conditional style that it's not assignet to it (but was assigned to other node in the script)

This happens with script filter conditioned styles.

when unfolding/folding its parent node the format is not refreshed and  stays wrong.

when selecting the node, its format gets refreshed and shows right.

**Comment by EdoFro** [comment]
Here is an example map with 3 test scripts:

[Test-conditionalStyles_createChild.zip](https://github.com/freeplane/freeplane/files/10069948/Test-conditionalStyles_createChild.zip)

instructions in the mindmap

**Comment by EdoFro** [comment]
## Second side effect (very similar):

when pasting a branch as clones, the new nodes don't get their conditional styles (script filters)

to reproduce:

1. use the example map form previous post
2. copy the conditional styles example branch
3. and paste it elsewhere as a clone

**Comment by EdoFro** [comment]
### _side effects_

Both side effects has been **solved** in **FP v1.10.6-pre02**.

I installed that version and **I will continue testing** to see if there are any other side effect left.

----

### _speed test_

I applied the same speed test to the new version too. Here are the results (in seconds):

|Test|Conditional Styles|1.10.5_03|1.10.5 stable|1.10.6_01|**1.10.6_02**|
|:--:|--|--:|--:|--:|--:|
|1|All (but "For any descendant..")|1,8|52,1|3,6|**3,6**|
|2|None|0,7|0,8|0,5|**0,7**|
|3|No script filters|0,9|3,1|3|**3,1**|
|4|No `(node.findAll() - node).any{...` filters|1,2|5,5|3,1|**3,3**|
|5|No groovyNode filters|1,3|16,4|3,2|**3,4**|
|6|only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: Script filter (node.fidAll() ….|1,1|28,9|0,6|**0,7**|
|7|only GroovyNode and  hasGroovyNode styles<br/>hasGroovyNode: For any descendant ….|2,3|291,3|2,3|**2,3**|
|8|All (but "For any descendant..") & without MapOverview|1,5|46,4|3,9|**3,5**|

It maintains its speed. It's a lot faster than the 1.10.5 stable version.

**Comment by EdoFro** [comment]
> I installed that version and **I will continue testing** to see if there are any other side effect left.

bug: when moving a node, it disappears

I created new bug report #844

I returned to version 1.10.6_01

**Comment by EdoFro** [comment]
> bug: when moving a node, it disappears

Bug solved in 1.10.6_03

I will continue testing to see if there are any other side effect left.

**Comment by EdoFro** [comment]
@dpolivaev 
I've been using version 1.10.6_03 all these days and I haven't found any other side effect.

I'll close this issue , so you can continue the process to publish a new stable version.

Thanks for all your great support!
